### ISSUE TITLE
Wrong type of decorated class

### ISSUE THREAD

**Issue by jer-sen** [microsoft/TypeScript#38971]
**TypeScript Version:**  3.9.2

**Search Terms:**

**Expected behavior:** No error

**Actual behavior:** Typescript doesn't use the result type of the decorator

**Related Issues:**

**Code**
```ts
const deco = (target: { new (...args: any[]): { [key in string]: any} & { originProp: boolean } }) => {
    return class NewClass extends target {
        newProp = true
    }
}

@deco
class OriginClass {
    originProp = true
}

const newInstance = new OriginClass();
newInstance.newProp; // Error : Property 'newProp' does not exist on type 'OriginClass'.

```

<details><summary><b>Output</b></summary>

```ts
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
const deco = (target) => {
    return class NewClass extends target {
        constructor() {
            super(...arguments);
            this.newProp = true;
        }
    };
};
let OriginClass = /** @class */ (() => {
    let OriginClass = class OriginClass {
        constructor() {
            this.originProp = true;
        }
    };
    OriginClass = __decorate([
        deco
    ], OriginClass);
    return OriginClass;
})();
const newInstance = new OriginClass();
newInstance.newProp; // Error : Property 'newProp' does not exist on type 'OriginClass'.

```

</details>

<details><summary><b>Compiler Options</b></summary>

```json
{
  "compilerOptions": {
    "noImplicitAny": true,
    "strictNullChecks": true,
    "strictFunctionTypes": true,
    "strictPropertyInitialization": true,
    "strictBindCallApply": true,
    "noImplicitThis": true,
    "noImplicitReturns": true,
    "useDefineForClassFields": false,
    "alwaysStrict": true,
    "allowUnreachableCode": false,
    "allowUnusedLabels": false,
    "downlevelIteration": false,
    "noEmitHelpers": false,
    "noLib": false,
    "noStrictGenericChecks": false,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "esModuleInterop": true,
    "preserveConstEnums": false,
    "removeComments": false,
    "skipLibCheck": false,
    "checkJs": false,
    "allowJs": false,
    "declaration": true,
    "experimentalDecorators": false,
    "emitDecoratorMetadata": false,
    "target": "ES2017",
    "module": "ESNext"
  }
}
```

</details>

**Playground Link:** [Provided](https://www.typescriptlang.org/play/?experimentalDecorators=true#code/MYewdgzgLgBAJgU1DAvDAFFAhgJwOYJQBcMA3jGAgO4YB09ueEJWYAngNoC6AlCeRwDWCNjACWYGNBwS8XFuwC+MAGRkYIGXgkAFHCAAOJAEYgQAGwSsYyxT1QA+MgFgAUDA8wchAK45JwOZYEBAwAHLUAMJBITAIAB5QCGBwodj4hC7unjmUVHqGqDBQOD4IbjmKblWubgACiKBugcGhAPJaEtGtWTmaYtpgBQZFJWXVbs3g0BTUAJKQ2GDACEV5MB0DXTEQ6DwA3G55C9CsK7R5w-swAPQ3MACiOPo4MCTDCDhQogDkl-oGH7wEAIUJgECwBJiGbgYpsAyrH6bQbdEI-WhuIA)

**Comment by MartinJohns** [comment]
Duplicate of #4881.
